library/lib.sh

```diff
@@ -260,20 +270,54 @@ rolesRunPlaybook() {
     local test_playbook=$2
     local inventory=$3
     local skip_tags=$4
+    local limit=$5
     local LOGFILE="${test_playbook%.*}"-ANSIBLE-"$ANSIBLE_VER"
     local result=FAIL
+    local cmd log_msg
+    cmd="ansible-playbook -i $inventory $skip_tags $limit $tests_path$test_playbook -vv"
+    log_msg="Test $test_playbook with ANSIBLE-$ANSIBLE_VER on ${limit/--limit /}"
     # If LSR_TFT_DEBUG is true, print output to terminal
     if [ "$LSR_TFT_DEBUG" == true ] || [ "$LSR_TFT_DEBUG" == True ]; then
-        rlRun "ANSIBLE_LOG_PATH=$LOGFILE ansible-playbook -i $inventory $skip_tags $tests_path$test_playbook -vv && result=SUCCESS" 0 "Test $test_playbook with ANSIBLE-$ANSIBLE_VER"
+        rlRun "ANSIBLE_LOG_PATH=$LOGFILE $cmd && result=SUCCESS" 0 "$log_msg"
     else
-        rlRun "ansible-playbook -i $inventory $skip_tags $tests_path$test_playbook -vv &> $LOGFILE && result=SUCCESS" 0 "Test $test_playbook with ANSIBLE-$ANSIBLE_VER"
+        rlRun "$cmd &> $LOGFILE && result=SUCCESS" 0 "$log_msg"
     fi
     logfile_name=$LOGFILE-$result.log
     mv "$LOGFILE" "$logfile_name"
     LOGFILE=$logfile_name
     rolesUploadLogs "$LOGFILE"
 }
 
+rolesRunPlaybooksParallel() {
+    # Run playbooks on managed nodes one by one
+    # Supports running against a single node too
+    local tests_path=$1
+    local inventory=$2
+    local skip_tags=$3
+    local test_playbooks=$4
+    local managed_nodes=$5
+
+    for test_playbook in $test_playbooks; do
+        while true; do
+            for managed_node in $managed_nodes; do
+                if ! pgrep -af "ansible-playbook" | grep -q "\-\-limit $managed_node"; then
+                    rolesRunPlaybook "$tests_path" "$test_playbook" "$inventory" "$skip_tags" "--limit $managed_node" &
```

Comment:
This is a valid concern. What you are saying relates to running ansible command like `ANSIBLE_LOG_PATH=$LOGFILE $cmd && result=SUCCESS"`, which is my mistake and should not be supported in parallel scenario. It's when `"$LSR_TFT_DEBUG" == true`, I added this for running a single test playbook just so that I can see playbook's output right away on my screen.

I was testing this with ansible command run as `$cmd &> $LOGFILE && result=SUCCESS`, the result is as is:
```
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::   Test
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

:: [ 08:58:08 ] :: [  BEGIN   ] :: Test tests_default.yml with ANSIBLE-2.17 on managed_node2 :: actually running 'ansible-playbook -i /var/tmp/tmt/run-159/plans/general/tree/postfix/inventory.yml --skip-tags tests::nvme,tests::infiniband --limit managed_node2 /tmp/tmp.UQVmkuPYdr/ansible_collections/fedora/linux_system_roles/tests/postfix/tests_default.yml -vv &> tests_default-ANSIBLE-2.17 && result=SUCCESS'
:: [ 08:58:13 ] :: [  BEGIN   ] :: Test tests_previous_replaced.yml with ANSIBLE-2.17 on managed_node1 :: actually running 'ansible-playbook -i /var/tmp/tmt/run-159/plans/general/tree/postfix/inventory.yml --skip-tags tests::nvme,tests::infiniband --limit managed_node1 /tmp/tmp.UQVmkuPYdr/ansible_collections/fedora/linux_system_roles/tests/postfix/tests_previous_replaced.yml -vv &> tests_previous_replaced-ANSIBLE-2.17 && result=SUCCESS'
:: [ 09:00:27 ] :: [   PASS   ] :: Test tests_default.yml with ANSIBLE-2.17 on managed_node2 (Expected 0, got 0)
:: [ 09:00:27 ] :: [   INFO   ] :: Sending tests_default-ANSIBLE-2.17-SUCCESS.log as tests_default-ANSIBLE-2.17-SUCCESS.log
File '/tmp/tmp.z3zcRGnTzC/tests_default-ANSIBLE-2.17-SUCCESS.log' stored to '/var/tmp/tmt/run-159/plans/general/execute/data/guest/control_node/Run-test-playbooks-from-control_node/tests/general-4/data'.
:: [ 09:00:28 ] :: [  BEGIN   ] :: Test tests_set_banner.yml with ANSIBLE-2.17 on managed_node2 :: actually running 'ansible-playbook -i /var/tmp/tmt/run-159/plans/general/tree/postfix/inventory.yml --skip-tags tests::nvme,tests::infiniband --limit managed_node2 /tmp/tmp.UQVmkuPYdr/ansible_collections/fedora/linux_system_roles/tests/postfix/tests_set_banner.yml -vv &> tests_set_banner-ANSIBLE-2.17 && result=SUCCESS'
:: [ 09:01:21 ] :: [   PASS   ] :: Test tests_previous_replaced.yml with ANSIBLE-2.17 on managed_node1 (Expected 0, got 0)
:: [ 09:01:21 ] :: [   INFO   ] :: Sending tests_previous_replaced-ANSIBLE-2.17-SUCCESS.log as tests_previous_replaced-ANSIBLE-2.17-SUCCESS.log
File '/tmp/tmp.YNLKyp1ZDS/tests_previous_replaced-ANSIBLE-2.17-SUCCESS.log' stored to '/var/tmp/tmt/run-159/plans/general/execute/data/guest/control_node/Run-test-playbooks-from-control_node/tests/general-4/data'.
:: [ 09:01:22 ] :: [  BEGIN   ] :: Test tests_set_file.yml with ANSIBLE-2.17 on managed_node1 :: actually running 'ansible-playbook -i /var/tmp/tmt/run-159/plans/general/tree/postfix/inventory.yml --skip-tags tests::nvme,tests::infiniband --limit managed_node1 /tmp/tmp.UQVmkuPYdr/ansible_collections/fedora/linux_system_roles/tests/postfix/tests_set_file.yml -vv &> tests_set_file-ANSIBLE-2.17 && result=SUCCESS'
:: [ 09:01:29 ] :: [   PASS   ] :: Test tests_set_banner.yml with ANSIBLE-2.17 on managed_node2 (Expected 0, got 0)
:: [ 09:01:29 ] :: [   INFO   ] :: Sending tests_set_banner-ANSIBLE-2.17-SUCCESS.log as tests_set_banner-ANSIBLE-2.17-SUCCESS.log
File '/tmp/tmp.0IkhXQ6igw/tests_set_banner-ANSIBLE-2.17-SUCCESS.log' stored to '/var/tmp/tmt/run-159/plans/general/execute/data/guest/control_node/Run-test-playbooks-from-control_node/tests/general-4/data'.
:: [ 09:01:51 ] :: [   PASS   ] :: Test tests_set_file.yml with ANSIBLE-2.17 on managed_node1 (Expected 0, got 0)
:: [ 09:01:51 ] :: [   INFO   ] :: Sending tests_set_file-ANSIBLE-2.17-SUCCESS.log as tests_set_file-ANSIBLE-2.17-SUCCESS.log
File '/tmp/tmp.D3AKvJHRC7/tests_set_file-ANSIBLE-2.17-SUCCESS.log' stored to '/var/tmp/tmt/run-159/plans/general/execute/data/guest/control_node/Run-test-playbooks-from-control_node/tests/general-4/data'.
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::   Duration: 224s
```

Log files are stored locally, or in the case of upstream tests, where env vars for ARTIFACTS_ and LINUXSYSTEMROLES_ are provided, stored in our Fedora storage.

Reply:
> > Terminal breaks after Running 'rlImport library' line
> 
> not sure what you mean by "breaks" - note that once you start running jobs in the background, the output from each background job, and the foreground job, will all be jumbled up together.

Oh, and you were referring to the issue that I opened for tmt, it appears in the initial phase of test run and cannot be related to this.